File: packages/python/src/mpesa/services/stk_push.py
from typing import Optional
from .http.http_client import HttpClient
from .auth import Auth
from .utils import Helpers


class STKPushService:
    def __init__(self, http_client: HttpClient, auth: Auth, config):
        self.http_client = http_client
        self.auth = auth
        self.helpers = Helpers(config)
        self.base_url = http_client.get_base_url()

    def push(self, phonenumber: str, amount: int, account_number: str, 
             callback_url: Optional[str] = None, transaction_type: str = 'CustomerPayBillOnline',
             short_code_type: str = 'C2B') -> dict:
        if not account_number:
            raise ValueError('An Account Reference is required for All transactions.')
        
        if transaction_type == 'CustomerBuyGoodsOnline' and not self.helpers.get_config('till_number'):
            raise ValueError('Till number is required for Buy Goods transactions.')
        
        url = f'{self.base_url}/mpesa/stkpush/v1/processrequest'
        
        data = {
            'BusinessShortCode': self.helpers.get_config('shortcode'),
            'Password': self.helpers.lipa_na_mpesa_password(),
            'Timestamp': self.helpers.get_formatted_timestamp(),
            'Amount': amount,
            'PartyA': self.helpers.phone_validator(phonenumber),
            'PartyB': self.helpers.get_config('shortcode') if transaction_type == 'CustomerPayBillOnline' else self.helpers.get_config('till_number'),
            'TransactionType': transaction_type,
            'PhoneNumber': self.helpers.phone_validator(phonenumber),
            'TransactionDesc': 'Payment',
            'AccountReference': account_number,
            'CallBackURL': callback_url or self.helpers.get_config('callbacks.callback_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, data, token)

    def query(self, checkout_request_id: str, short_code_type: str = 'C2B') -> dict:
        url = f'{self.base_url}/mpesa/stkpushquery/v1/query'
        
        data = {
            'BusinessShortCode': self.helpers.get_config('shortcode'),
            'Password': self.helpers.lipa_na_mpesa_password(),
            'Timestamp': self.helpers.get_formatted_timestamp(),
            'CheckoutRequestID': checkout_request_id,
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, data, token)

File: packages/node/src/services/stkPush.ts
import { StkPushParams } from '../types';
import { HttpClient } from '../http/httpClient';
import { Auth } from '../core/Auth';
import { Helpers } from '../utils/helpers';

export class STKPushService {
  private httpClient: HttpClient;
  private auth: Auth;
  private helpers: Helpers;
  private baseUrl: string;

  constructor(httpClient: HttpClient, auth: Auth) {
    this.httpClient = httpClient;
    this.auth = auth;
    this.helpers = new Helpers();
    this.baseUrl = httpClient.getBaseUrl();
  }

  async push(params: StkPushParams): Promise<any> {
    const { phonenumber, amount, accountNumber, callbackUrl, transactionType = 'CustomerPayBillOnline', shortCodeType = 'C2B' } = params;

    if (!accountNumber) {
      throw new Error('An Account Reference is required for All transactions.');
    }

    const url = `${this.baseUrl}/mpesa/stkpush/v1/processrequest`;
    const data = {
      BusinessShortCode: this.helpers.getConfig('shortcode'),
      Password: this.helpers.lipaNaMpesaPassword(),
      Timestamp: this.helpers.getFormattedTimestamp(),
      Amount: amount,
      PartyA: this.helpers.phoneValidator(phonenumber),
      PartyB: transactionType === 'CustomerPayBillOnline' ? this.helpers.getConfig('shortcode') : this.helpers.getConfig('tillNumber'),
      TransactionType: transactionType,
      PhoneNumber: this.helpers.phoneValidator(phonenumber),
      TransactionDesc: 'Payment',
      AccountReference: accountNumber,
      CallBackURL: callbackUrl || this.helpers.getConfig('callbacks.callbackUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, data, token);
  }

  async query(checkoutRequestId: string, shortCodeType: string = 'C2B'): Promise<any> {
    const url = `${this.baseUrl}/mpesa/stkpushquery/v1/query`;
    const data = {
      BusinessShortCode: this.helpers.getConfig('shortcode'),
      Password: this.helpers.lipaNaMpesaPassword(),
      Timestamp: this.helpers.getFormattedTimestamp(),
      CheckoutRequestID: checkoutRequestId,
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, data, token);
  }
}
